### DISCUSSION TITLE
ZFS kernel threads priority

### DISCUSSION THREAD

**Discussion by slavanap** [openzfs/zfs#14258] [General]
I'm looking for a setting that would assist in resolution of the following issue:
1. create zfs dataset with compression=gzip-9
2. start audio playback with pulseaudio & vlc (or similar)
3. copy 10GB worth of data to newly created dataset
4. observe as audio playback hangs until copying is completed

Is there any way to operate with huge amounts of data stored in ZFS and play audio without interruptions simultaneously?

I tried chrt vlc & pulseaudio -- it does not help. Probably USB driver is missing CPU cycles, because ZFS with compression=gzip-9 loads 100% of all CPU cores with kernel threads for compression. ZFS kernel threads nice value is -19 with SCHED_OTHER. They are spawned frequently, thus I can't alter their priority consistently. Maybe there's a way to drop priority change for them or spawn them in SCHED_IDLE queue temporarily?

OS: Ubuntu 22.04.1, CPU: Intel Core i7, 2 cores

**Comment by FFourtyTwo** [comment]
Hello!
First of all, why gzip-9? By my experience this is one of most ineffective and intensive algorithms for any workload. Also why so high ratio "-9"? For your 2c/4t CPU it looks like too much.

I suggest you to try zstd algorithm first, cause it's most effective we have starting from ZFS 2.0. If you really need to save every byte you can try zstd-19 but remember that every ratio step is less efficient than previous one. Depending on your data it may not worth to set high ratios. For example you will get +0.5% of space but it will take +50% of CPU time.

By my opinion default rate of compression=zstd-3 (that equals to compression=zstd) is best option for generic workloads. If you wanna save more than use zstd-5, zstd-7. After this (zstd-9+) it will not worth it.

Check this:
https://www.reddit.com/r/zfs/comments/svnycx/a_simple_real_world_zfs_compression_speed_an/
https://freebsdfoundation.org/wp-content/uploads/2021/05/Zstandard-Compression-in-OpenZFS.pdf
https://openzfs.org/w/images/4/4d/Compression-Saso_Kiselkov.pdf

Don't forget that any process scheduler (CFS in linux) works the worse the higher the load. In your case with 2c/4t and 100% load your scheduler spends much time to switch execution context. I suggest first to lower the overall load using better compression algorithm.

**Reply by slavanap** [reply]
Gzip-9 is because I use the dataset for archive purposes and gzip provides 60% compression ratio vs. 20-25% of zstd. The whole point of the question is **not** about it. Once compression threads start to load CPU because disk bandwidth is higher than compressor throughput the issue with the audio will occur with any compression algorithm.

> Don't forget that any process scheduler (CFS in linux) 

I use SCHED queues and load CPU with no issues at 100% while audio is playing and long running tasks are being running consuming remaining 100% of available CPU idle time. I experience issues only with ZFS now because it runs at high priority and steals CPU cycles from processes from which it should not to.

The topic question is that is there currently an option to alter ZFS threads priority so they won't interfere with other kernel threads operation as they are now at high throughput IO?

**Reply by IvanVolosyuk** [reply]
Are you confusing lz4 and zstd? I think zstd-9 supposed to have similar compression level as gzip-9.

**Comment by slavanap** [comment]
https://github.com/openzfs/zfs/commit/1229323d5f82bcb0525b312c80330968cd41466e

**Comment by IvanVolosyuk** [comment]
I had similar issues with sound in my qemu/kvm workload. I managed to fully mitigate it by using preemptive kernel  (```CONFIG_PREEMPT=y```) and also setting ```spl.spl_taskq_thread_bind=0 zfs.spl_taskq_thread_bind=0``` in kernel boot parameters. One of them is unused but I didn't spend time finding out which one.

My theory is system devices like sound or network device use cpu affinity mask to operate on a specific CPU thread, but ZFS will also schedule big chunk of work on those CPU thread which cannot be reallocated to another one. Also, compression in ZFS without preemptive kernel doesn't let higher priority tasks to be executed for a long time even if they are latency critical.

I only had this issue with qemu and I run it with realtime scheduling priority using ```chrt -r 5 qemu...```. This command also allows to adjust scheduling for already running processes. You can try to bump up priority of pulseaudio and your sound app.

**Reply by Zocker1999NET** [reply]
Most probably pulseaudio / pipewire has already a higher, preferred priority (if not even being configured to be a realtime process) to avoid such problems as this should be the default for most graphical installations.

**Comment by Zocker1999NET** [comment]
I had a similar problem but with a much more capable setup and less performance dragging configuration: I only had a few audio stutters but my CPU (Ryzen 4 1600, 6 cores / 12 threads) was only about 50 % max due to ZFS compression. Normally, I could utilize my full CPU power (e.g. compiling Lineage OS with 12 threads) and I wouldn't hear any audio stutters due to working priorities and scheduling. ZFS compression seem to be excluded from that.

I could mitigate them for me without rebooting while restarting mpv with `PIPEWIRE_LATENCY=1024/48000 mpv` (increasing pipewire's buffers).
Next boot I may try the kernel boot parameters.

But still I would wish to be able to deprioritize ZFS's compression thread properly, best with an option possible to alter while the system is running.

**Reply by CaptaiNiveau** [reply]
How'd that go? I'm getting some audio stutters on encrypted zfs root with full preemption on a 5800x. LZ4 compression shouldn't be an issue, audio stutters are also quite rare. Haven't been able to find a simple way to reproduce the issue, but running fzf with fd in a big folder seems to trigger it.

**Comment by memeruiz** [comment]
I had a similar problem. Also with a much more capable system: Ryzen 9 5900X 24-threads 94GB RAM. Using zstd-10 copying from from zfs (default compression algo) to a newer zfs (zstd-10).

First problem I experimented was, slow copy speed. Checking I noticed a lot of swap activity. I manage to "control" this by changing a couple of zfs parameters concerning memory usage. The first was zfs_arc_max. I set it to much less than 1/2 of RAM. The other was zfs_arc_sys_free. First tried to set it to 10GB and it didn´t solve the problem, 40GB did it. My guess is that zstd in zfs using a separated RAM space which depends on the zstd chosen level, with higher requiring more RAM. Since zfs arc RAM usage has apparently nothing to do with zstd then zfs arc RAM was strangling zstd and sending it to swap. 

Once this problem was solved then during copy, with some files CPU usage went really somewhat high, using basically all CPU threads and causing the same type of mpv audio problems. I played with ionice and renice, giving the rsync (copy) processes the lowest possible priority and mpv the highest possible priority but it didn't improve the audio problem. Also I tried the PIPEWIRE_LATENCY=1024/48000 suggested above and it didn't improve audio glitching. spl.spl_taskq_thread_bind=0 was already to 0.

There must be some zstd related kernel/module activity that has a really high priority that can't be changed. 

Computer is completely unusable while zfs compress copy. Not only mpv audio is affected, browsing and everything else is sluggish. A Ryzen 9 brought to its knees with just this.

I think the best solution would be to have a more comprehensive memory control of zfs (not just arc) that includes memory control for all zfs activities, and also a priority setting for different zfs activities (hashing, encryption and compression). 

Edit: The latency problem was solved changing preempt from voluntary to full:

echo "full" | sudo tee /sys/kernel/debug/sched/preempt

I thought full was the default with Debian Kernels

In short. To use zstd above 3 one needs to 
- Reduce zfs arc memory usage (to avoid swap usage)
- Increase zfs arc sys free parameter (to avoid swap usage)
- Switch preempt to full (to avoid sluggishness) 

Kernel 6.4.0-0-amd64
zfs version: 2.1.12-1

**Reply by rincebrain** [reply]
any zstd above 9 is an incredibly bad idea in general, and i wouldn't go above 6 for almost any use case.

maybe i'll go open a PR disabling your ability to do that without a tunable since so many people keep thinking bigger is always better and being surprised it runs like a steaming pile.

**Reply by slavanap** [reply]
The issue is not about what the compression ratio is. The issue is about ZFS developers keep thinking that their software can consume all CPU anytime their software wants, and that's why ZFS kernel threads have the highest priority. Maybe I'll go open PR that defaults their priority back to normal alongside audio and video output threads.

**Reply by IvanVolosyuk** [reply]
Do you use preemtive kernel (```CONFIG_PREEMPT=y```)?
I believe that there is not enough preemption points in zfs's zstd compression code which causes starvation of other processes even with really high priority as they have to wait for a zthread to yield voluntarily. The only solution I found for myself is fully preemptive kernel, which works pretty well.

Also, zstd code allocates per-thread buffers it seems with the size depending on zstd compression level, thus for CPUs with many hardware threads zfs spawns many zthreads consuming a lot of ram, which is not accounted by arc tunables.

**Reply by rincebrain** [reply]
Open a feature request, or just add onto e.g. #15142 requesting other thread changes, but the reason the threads were set to that priority, IIRC, was that, if you try looking, the Linux kernel sets a lot (all?) of its IO-related threads to -20. Go look at what all the XFS kernel threads run as, for example.

So setting it to something lower would enable fun priority inversion problems where lower priority things can block -20 IO threads.

You can still do it - there's a tunable to stop setting priority on threads, but I don't think you'll like the outcome.

Also, the per-hwthread buffers should be only as large as they ever needed to be to satisfy a request, so if they're getting large for you, then it's likely because they satisfied large requests. I'm not sure what a better pooling method would be that wouldn't have weird memory locality effects than "one per hwthread", though...

**Reply by memeruiz** [reply]
Already with zstd-5 it was horrible.

**Reply by memeruiz** [reply]
@IvanVolosyuk I just configured to preempt=full and now all is fine. I had it at "voluntary". No more sluggishness. Did it before seeing your comment, but thanks! Though of it a bit late! To be honest, I thought debian kernels came preempt=full by default, but no.

**Reply by slavanap** [reply]
@rincebrain, encryption and compression in ZFS is not I/O, but its priority is on I/O level.

**Reply by rincebrain** [reply]
I'm not really sure how you think "bytes being read/written to disk need to be {de,}{compressed,encrypted}" isn't in the IO pipeline, but you're wrong.

**Reply by Zocker1999NET** [reply]
> So setting it to something lower would enable fun priority inversion problems where lower priority things can block -20 IO threads.

With that perspective, it makes sense to me why ZFS compression threads have a high priority as ZFS has no idea if its current tasks are essential to the system (being accesses to the system drive) or not (copying files sequentially to external hard drives).

Might it be possible to allow users selecting a priority for those ZFS compression threads *per pool* (maybe even per dataset would be nice) and *without re-compiling the kernel or the ZFS module*?
If setting kernel thread priorities dynamically is not really possible or feasible, it could be enough if there are two thread groups for compression/encryption/IO, one with -20 (still being the default) & one with a much lower priority. With an option per pool/dataset, admins can then select how important access to these pools/datasets are while docs hint them at the risks that come with deprioritizing compression/encryption.

An example use case is to have a system and an archival pool. The system pool is allowed to use the high priority threads and will also only have zstd compression as it is vital to the system. The archival pool should use the non-prioritized IO threads for best compression possible, as 1st, it is not vital to the system and its programs, 2nd, the higher compression rate is more valued than access latency and, 3rd, the system is expected to have enough free time for that lower priority IO thread, and if not, then there are other issues more important to the users (graphics, sound, input latency, …).

**Comment by MagicRB** [comment]
I'm having a issue from a similar bucket. I use zfs on my workstations and while it works great, what i've noticed is that during heavy IO workloads pipewire will not get enough cpu time and then the audio will crackle for a moment. I'm already running with `preempt=full`, `zstd-2` globally, i7-12700H, raid1 over 2 samsung 980 pros. So i've done all the usual things. 

What concerns me, is that ZFS has higher priority than pipewire, even though id consider pipewire to be a more high priority task personally. I've been considering to invert the priority of pipewire and zfs, pipewire is currently on -11 and zfs on -19 (at least the write threads). so id need to at least move zfs to -18 and pipewire to -19 which seems extreme.

Not sure where to take this from here, but it's been an issue for years.

**Reply by kotauskas** [reply]
PipeWire's nice -11 applies only to the main thread. You can see in `htop` if you display user threads with Shift+H and sort by priority descending that the data loop threads really have higher priority than nice -19 processes. Here's how this looks on my system:
<img width="1382" height="313" alt="htop screenshot" src="https://github.com/user-attachments/assets/2aeff3ca-0b58-4d13-8460-f50118c42ccc" />

**Reply by MagicRB** [reply]
Main thread or not, pipewire being less prioritized than zfs seems wrong. Idk if thats the cause of my stutter, but i know the stutter corellates zfs handing writes.

**Reply by amotin** [reply]
I am not sure kernel threads of ZFS can have priority lower than user-space. Though it is OS-specific.  What you may try is to slightly reduce a number of write threads via `zio_taskq_batch_pct`.

**Reply by kotauskas** [reply]
> Main thread or not, pipewire being less prioritized than zfs seems wrong.

It is not wrong, as the main thread of PipeWire does not contain the audio processing loop and does not need to have high priority. It could be nice 20 and it wouldn't result in stuttering in the middle of playback. The worst that can happen is that starting and stopping playback gets laggy under high CPU load. If PipeWire's main thread had real-time work on it (the kind of work that leads to crackling if delayed), it would be real-time just like the data loop threads.

> Idk if thats the cause of my stutter, but i know the stutter corellates zfs handing writes.

You can test this by doing `sudo renice -n-19 $(pgrep pipewire)`. If the crackling remains (as it most likely will), the problem is not ZFS threads preempting PipeWire's main thread, but some other thread (naturally, one of the threads in the process that is playing audio) having insufficient priority.

The best way to diagnose this is to crack open `htop`, add the `SCHED` and `PRIORITY` columns, show user threads with Shift+H, sort by `PRIORITY` descending, and take a look at the threads that float above the priority 0 mark. If the process playing audio isn't on that list, you have an RTKit-related issue or the issue of that particular program not setting a real-time scheduling policy for its playback threads.

> I am not sure kernel threads of ZFS can have priority lower than user-space.

At least on Linux, they definitely can, at least on account on all the ZFS kthreads with nice 20 which do not seem to preempt anything (as intended). Additionally, in my experience of getting audio stuttering from ZFS I/O, switching from `preempt=voluntary` to `preempt=full` fixed the problem, which implies that the kthreads are being preempted by userspace work as one would expect.

**Reply by amotin** [reply]
I don't remember it on Linux, but at least on FreeBSD kernel threads just can't be time-share, they always have static priority.

**Reply by IvanVolosyuk** [reply]
I also use 'spl.spl_taskq_thread_bind=0 spl.spl_taskq_thread_priority=0' kernel parameters and don't use zstd, just lz4. I think the problem with ZFS compression is that it can occupy a CPU thread for a long time and if the CPU has affinity to handle sound card interrupts - it can cause extra latency. Making ZFS threads migratable can potentially cause ZFS performance decrease, but will let that CPU thread handle the sound card better.

**Reply by kotauskas** [reply]
It would be a little strange if `spl.spl_taskq_thread_bind=0` affected audio stuttering on `preempt=full`, since the sound card interrupt related work would obviously preempt the taskq thread. Additionally, `spl.spl_taskq_thread_priority=0` can only possibly help if the scheduling policies of audio playback data loop threads are not `SCHED_RR`, which is a userspace misconfiguration of the audio system that will cause audio stuttering under any kind of high CPU load, rather than a problem with ZFS in specific.

**Reply by slavanap** [reply]
If you want to solve the audio issue, you may consider changing -20 and -19 here to 0 on your setup https://github.com/openzfs/zfs/blob/master/lib/libspl/include/sys/sysmacros.h#L118
```
#define wtqclsyspri     -19
#define maxclsyspri     -20
```
to
```
#define wtqclsyspri     0
#define maxclsyspri     0
```

ZFS developers set priority for its kernel threads to maximum which steals CPU cycles from all other threads when ZFS IO is busy with compression/encryption. The priority change started from the commit I referenced in my previous message.

As I grasp from other messages by the ZFS team they position ZFS IO as a first class citizen in Linux kernel, primarily because ZFS goal is to run as fast as possible on file servers, and they have no plans to change current ZFS behavior themselves.

**Reply by Zocker1999NET** [reply]
> If you want to solve the audio issue, you may consider changing -20 and -19 here to 0 on your setup https://github.com/openzfs/zfs/blob/master/lib/libspl/include/sys/sysmacros.h#L118

Thanks, I will try that out.

> As I grasp from other messages by the ZFS team they position ZFS IO as a first class citizen in Linux kernel, primarily because ZFS goal is to run as fast as possible on file servers, and they have no plans to change current ZFS behavior themselves.

I mean for the intended usecase of ZFS, this is a somewhat sensible default. But then I have question to the ZFS devs: would it be possible to make those two numbers configurable via syscfg / kernel parameters? At least at boot without the capability to change them during runtime. And if wanted with a documented warning: "performance may drop significantly / system may effectively hang / you are on your own / ...". I would also accept a warning in dmesg logged at every boot warning about this.

Because this then would be great for people using ZFS on their desktop system for reasons other than performance (snapshotting, send&recv compatibility with NAS, ...) to find a common ground without the need to recompile the ZFS kernel module just to change two constants.

**Reply by IvanVolosyuk** [reply]
The kernel parameter is `spl.spl_taskq_thread_priority=0`.
